src/ts/contentScripts/live-preview/index.tsx

```diff
@@ -0,0 +1,293 @@
+import {Feature, Settings} from '../../utils/settings'
+import {Navigation} from '../../roam/navigation'
+import {browser} from 'webextension-polyfill-ts'
+import {createPopper, Instance} from '@popperjs/core'
+
+export const config: Feature = {
+    id: 'live_preview',
+    name: 'Live Preview',
+    warning: 'Experimental feature; Large databases might see performance issues.',
+    enabledByDefault: false,
+}
+const checkSettingsAndSetupIframeToggle = () => {
+    Settings.isActive('live_preview').then(active => {
+        toggleIframe(active)
+    })
+}
+checkSettingsAndSetupIframeToggle()
+
+browser.runtime.onMessage.addListener(async message => {
+    if (message === 'settings-updated') {
+        checkSettingsAndSetupIframeToggle()
+    }
+})
+
+let iframeInstance: PreviewIframe | null = null
+
+const toggleIframe = (active: boolean) => {
+    if (!iframeInstance) {
+        iframeInstance = new PreviewIframe()
+    }
+    if (active) {
+        iframeInstance.activate()
+    } else {
+        iframeInstance.destroy()
+        iframeInstance = null
+    }
+}
+
+class PreviewIframe {
+    iframeId = 'roam-toolkit-iframe-preview'
+    iframe: HTMLIFrameElement | null = null
+    popupTimeout: ReturnType<typeof setTimeout> | null = null
+    hoveredElement: HTMLElement | null = null
+    popper: Instance | null = null
+    popupTimeoutDuration = 300
+    activate() {
+        this.initPreviewIframe()
+    }
+    destroy() {
+        this.removeIframe()
+        this.clearPopupTimeout()
+        this.destroyPopper()
+        this.destoryMouseListeners()
+    }
+    private clearPopupTimeout() {
+        if (this.popupTimeout) {
+            clearTimeout(this.popupTimeout)
+            this.popupTimeout = null
+        }
+    }
+
+    private removeIframe() {
+        const isCurrentIframePresent = document.body.contains(this.iframe)
+        if (!this.iframe || !isCurrentIframePresent) {
+            return
+        }
+        if (isCurrentIframePresent) {
+            document.body.removeChild(this.iframe)
+        }
+        if (this.iframe) {
+            this.iframe = null
+        }
+    }
+
+    private getIFrameByUrl(url: string): HTMLIFrameElement | null {
+        return document.querySelector(`iframe[src="${url}"]`)
+    }
```

Comment:
@Stvad Haha, I wish 😅 It is not possible in prettier because of the way prettier works.See this more more info
> It turns out that empty lines are very hard to automatically generate. The approach that Prettier takes is to preserve empty lines the way they were in the original source code
https://prettier.io/docs/en/rationale.html#empty-lines

Reply:
huh, interesting. May I suggest you configure your IDE to do so then ?😛